server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/ImageUploadFieldWidget.java

```diff
@@ -0,0 +1,311 @@
+                uploadFailed.setVisible(true);
+            }
+
+            @Override
+            public void upload() {
+                ImageUploadFieldWidget.this.upload();
+            }
+        });
+        fileUpload.addChangeHandler(new ChangeHandler() {
+            @Override
+            public void onChange(ChangeEvent event) {
+                if (fieldWidgetMode == FieldWidgetMode.NORMAL) {
+                    uploader.requestUploadUrl();
+                    highlight(false);
+                } else {
+                    Window.alert(I18N.CONSTANTS.uploadIsNotAllowedInDuringDesing());
+                }
+            }
+        });
+
+        addFileDnDSupport();
+    }
+
+    private void addFileDnDSupport() {
+        imageContainer.addDomHandler(new DragEnterHandler() {
+            @Override
+            public void onDragEnter(DragEnterEvent event) {
+                highlight(true);
+            }
+        }, DragEnterEvent.getType());
+
+        imageContainer.addDomHandler(new DragOverHandler() {
+            @Override
+            public void onDragOver(DragOverEvent event) {
+                highlight(true);
+            }
+        }, DragOverEvent.getType());
+
+        imageContainer.addDomHandler(new DragLeaveHandler() {
+            @Override
+            public void onDragLeave(DragLeaveEvent event) {
+                highlight(false);
+            }
+        }, DragLeaveEvent.getType());
+
+        imageContainer.addDomHandler(new DropHandler() {
+            @Override
+            public void onDrop(DropEvent event) {
+                //event.preventDefault();
+
+                // we are following the mouse with fileUpload which leads to dropping directly to input type=file
+            }
+        }, DropEvent.getType());
+    }
+
+    private void highlight(boolean enter) {
+        if (enter) {
+            imageContainer.addStyleName(FormPanelStyles.INSTANCE.dropFile());
+        } else {
+            imageContainer.removeStyleName(FormPanelStyles.INSTANCE.dropFile());
+        }
+
+        fileUpload.setVisible(enter);
+        setLoadingState(false);
+        image.setVisible(!enter);
+        clearButton.setVisible(false);
+    }
+
+    private void upload() {
+        GwtUtil.setVisible(false, downloadButton);
+        uploadFailed.setVisible(false);
+        setLoadingState(true);
+
+        if (oldHandler != null) {
+            oldHandler.removeHandler();
+        }
+
+        oldHandler = formPanel.addSubmitCompleteHandler(new FormPanel.SubmitCompleteHandler() {
+            @Override
+            public void onSubmitComplete(FormPanel.SubmitCompleteEvent event) {
+                // event.getResults is always null because of cross-domain upload
+                // we are forced to make additional call to check whether upload is successful
+
+                fetchImageServingUrl();
+            }
+        });
+        formPanel.submit();
+    }
+
+    public void fetchImageServingUrl() {
```

Comment:
I like idea of fetching more because then we can log problems and also give feedback to user (Upload failed message).